### PR TITLE
UPSTREAM: 44058: Make background garbage collection cascading

### LINKED ISSUE #14805: RC can end up with a Foreground finalizer but no deletionTimestamp
This is related to https://github.com/kubernetes/kubernetes/issues/47843
Except when it is a DC and not a Deployment, then on the RC that gets left behind we are seeing a finalizer get set but there is no deletionTimestamp on the RC.  It should not have a foreground deletion finalizer with no deletionTimestamp.  This may not be a separate issue but tracking it in case there is something weird going on in our deployment controller.

### PR DESCRIPTION
**Pull request by soltysh**
This is upstream cherry-pick and revert of the changes that happened in origin in the mean time to address problems with garbage collection.

Fixes #14805

@liggitt @jwforres @mfojtik fyi
@tnozicka ptal

### CODE REVIEW

**Comment by soltysh** [comment]
[test]

File: pkg/controller/controller_ref_manager.go [hunk 1]
```diff
@@ -142,12 +142,10 @@ func (m *RCControllerRefManager) AdoptReplicationController(rs *v1.ReplicationCo
 		`{"metadata":{
 			"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],
 			"uid":"%s",
```
Comment (tnozicka) [inline pkg/controller/controller_ref_manager.go]: @soltysh that trailing comma is invalid JSON

**Comment by tnozicka** [comment]
@liggitt I don't think we have backported the finalizers in Deployments yet so nothing to revert for Deplyoments, but I would have to check

**Comment by tnozicka** [comment]
https://github.com/openshift/origin/pull/14880 should go first to be sure about changing JSON and it also tests cascade deletion for DC->RC->Pod

File: pkg/controller/controller_ref_manager.go (hunk 1 above)
Comment (soltysh) [inline pkg/controller/controller_ref_manager.go]: Good catch, will fix.

**Comment by tnozicka** [comment]
@liggitt Deployment finalizers we not backported https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/controller/controller_ref_manager.go#L352-L352

**Comment by soltysh** [comment]
> Do we also need the upstream revert that did similar things?

@liggitt the bits that happened upstream were reverted as part of the same PR, iiuc. Nothing else was added.

**Comment by liggitt** [comment]
LGTM as well

**Comment by tnozicka** [comment]
https://github.com/openshift/origin/pull/14880 is in; [test]

**Comment by tnozicka** [comment]
The new tests succeeded. Ship it!

**Comment by soltysh** [comment]
[merge] on two lgtm

**Comment by deads2k** [comment]
Green tests against current HEAD. No conflicts with top of the queue. merging

**Pull request merged**